### ISSUE TITLE
Work on perf in gg_miss for large dataset

### ISSUE THREAD

**Issue by ColinFay** [njtierney/naniar#101]
gg_miss() family seems to take a (very) long time to compute on large dataset. 

Here's an example w/ `diamonds`: 
```
mb <- microbenchmark::microbenchmark(gg_miss_case(diamonds), times = 10)
mb
```
![capture d ecran 2017-08-24 a 14 52 04](https://user-images.githubusercontent.com/17936236/29719771-d3cd7244-89b7-11e7-84b3-df1e604877df.png)

Is this ggplot or naniar related?

**Comment by ColinFay** [comment]
Here's a quick shot at rewriting `gg_miss_case` in base R: 

```r
miss_case_summary_base <- function(data, order = FALSE, ...){
  
  res <- data
  res$pct_miss <- apply(X = data, MARGIN = 1, FUN = function(x) (mean(is.na(x)) * 100))
  res$n_miss <- apply(X = data, MARGIN = 1, FUN = function(x) (sum(is.na(x))))
  res$case <- 1:nrow(data)
  res$n_miss_cumsum <- cumsum(res$n_miss)
  res <- dplyr::select(res, 
                       case,
                       n_miss,
                       pct_miss,
                       n_miss_cumsum)
  
  if (order) {
    return(dplyr::arrange(res, -n_miss))
  } else {
    return(res)
  }
}

# Outputs are the same 

base_output <- miss_case_summary_base(airquality)
current_output <- miss_case_summary(airquality)

all_equal(base_output, current_output)
[1] TRUE

`````

And here's the results from a microbenchmark w/ airquality and diamonds:

```r
mb <- microbenchmark::microbenchmark(orig = miss_case_summary(airquality), 
                                     old_school = miss_case_summary_base(airquality))

Unit: milliseconds
       expr       min        lq     mean    median        uq       max neval
       orig 349.15314 393.77068 468.8828 423.84853 509.88387 937.31800   100
 old_school  21.25257  22.69959  28.5989  25.35215  32.29996  51.44304   100
```
The base solution is here ~16 times quicker than the current solution. But still, a matter of milliseconds here.

Things get more obvious when we move to diamonds : 

```r
library(ggplot2)
data(diamonds)
mb2 <- microbenchmark::microbenchmark(orig = miss_case_summary(diamonds), 
                                     old_school = miss_case_summary_bis(diamonds))

Unit: seconds
       expr       min         lq       mean     median         uq        max neval
       orig 67.256863 102.099246 102.850157 104.472683 106.681096 116.636314   100
 old_school  3.222446   3.634497   3.823108   3.770275   3.980973   5.133206   100
```

This is another order of magnitude :) We've got here units in seconds, and current `miss_case` is around 25 times slower than the function written in base. 

When we tidy a little with a mutate : 
```r
miss_case_summary_mut <- function(data, order = FALSE, ...){
  
  res <- data %>% 
    dplyr::mutate(pct_miss = apply(X = ., MARGIN = 1, FUN = function(x) (mean(is.na(x)) * 100)),
                  n_miss = apply(X = data, MARGIN = 1, FUN = function(x) (sum(is.na(x)))), 
                  case = 1:nrow(.), 
                  n_miss_cumsum = cumsum(n_miss)) %>%
    dplyr::select(case,
                  n_miss,
                  pct_miss,
                  n_miss_cumsum)
  
  if (order) {
    return(dplyr::arrange(res, -n_miss))
  } else {
    return(res)
  }
}

mb3 <- microbenchmark::microbenchmark(mut = miss_case_summary_mut(airquality), 
                                     old_school = miss_case_summary_base(airquality))

Unit: milliseconds
       expr      min       lq     mean   median       uq      max neval
        mut 18.96697 19.56027 20.89321 20.08280 20.99360 44.78389   100
 old_school 11.88736 12.35850 13.12145 12.70893 13.54379 18.63309   100

library(ggplot2)
data(diamonds)
mb4 <- microbenchmark::microbenchmark(mut = miss_case_summary_mut(diamonds), 
                                      old_school = miss_case_summary_base(diamonds))

Unit: seconds
       expr      min       lq     mean   median       uq      max neval
        mut 1.338729 2.427672 2.439025 2.532146 2.602633 2.772421   100
 old_school 1.264281 2.376645 2.347995 2.460165 2.546791 3.255949   100

```

The mutate solution is as quick as the base solution.

**Comment by njtierney** [comment]
:tada: @ColinFay Thanks for taking the time to do this, it is super awesome, great job! :)

You have hit upon all of the things that I wanted to know about: performance and equal output.

It is interesting that base operations really are much faster, good to know that we've probably got lots of room for improvement in the future.

How would you feel about wrapping this up into a pull request in the next week or two, in time for the CRAN release (v0.2.0 or v0.3.0 [pending on how much we get done!]) at the end of this month?

**Comment by ColinFay** [comment]
{purrrlyr} seems to be the biggest time consumer here : 

```r
miss_case_table.default <- function(data){
  
  purrrlyr::by_row(.d = data,
                   # how many are missing in each row?
                   ..f = ~n_miss(.),
                   .collate = "row",
                   .to = "n_miss_in_case") %>%
    dplyr::group_by(n_miss_in_case) %>%
    dplyr::tally() %>%
    dplyr::mutate(pct_miss = (n / nrow(data) * 100)) %>%
    dplyr::rename(n_cases = n)
  
}

a <- miss_case_table(airquality)

miss_case_table_base <- function(data){
  res <- data
  res$n_miss_in_case <- apply(data, MARGIN = 1, FUN = n_miss)
  res %>%
    dplyr::group_by(n_miss_in_case) %>%
    dplyr::tally() %>%
    dplyr::mutate(pct_miss = (n / nrow(data) * 100)) %>%
    dplyr::rename(n_cases = n)
  
}

b <- miss_case_table_base(airquality)
all_equal(a, b)
[1] TRUE

library(ggplot2)
data(diamonds)
mb_table <- microbenchmark::microbenchmark(base_table = miss_case_table_base(diamonds))

Unit: milliseconds
expr      min       lq     mean   median       uq      max neval
base_table 620.2758 661.8756 1065.734 719.7365 802.8198 5155.629   100
```

Just rewrote the {purrrlyr} part and this new version is ~20 times quicker than the actual.

Not sure rewriting the dplyr part is worth the shot though. 

I'll send a PR this weekend :)

**Comment by ColinFay** [comment]
Hey, 

Is this issue fixed by the cpp PR from Romain?

**Comment by njtierney** [comment]
Yes, ish! We're still merging the pull request at the moment actually - just doing some final testing. 

I'll close this once I have the final bit of rcpp code merged in, it would be nice to keep all of the code together so we can look at the speedups! :)

**Comment by njtierney** [comment]
OK so I couldn't quite reprex this as I would have liked (gist of the full doc is [here](https://gist.github.com/njtierney/5d1eec9bf9e79e1a14090c01b4302907#file-git-branch-speed-comparison-r-L9) - but here is a summary of the speed improvements: These will get merged in very soon.

``` r

# tidy up the benchmark output into a tibble
tidy_mb <- function(mb_obj){
  tibble::as_tibble(mb_obj) %>%
    dplyr::arrange(expr) %>%
    dplyr::group_by(expr) %>%
    dplyr::mutate(row_id = 1:n()) %>%
    dplyr::ungroup()
}

mb_old_funs <- readr::read_rds("~/Downloads/mb_old_master.rds")
mb_new_funs <- readr::read_rds("~/Downloads/mb_new_row_means.rds")

library(tidyverse)
#> ── Attaching packages ─────────────────────────────────────────────────────────────────────────── tidyverse 1.2.1 ──
#> ✔ ggplot2 2.2.1.9000     ✔ purrr   0.2.4     
#> ✔ tibble  1.3.4          ✔ dplyr   0.7.4     
#> ✔ tidyr   0.7.2          ✔ stringr 1.2.0     
#> ✔ readr   1.1.1          ✔ forcats 0.2.0
#> ── Conflicts ────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
#> ✖ dplyr::filter() masks stats::filter()
#> ✖ dplyr::lag()    masks stats::lag()

tidy_mb_old <- tidy_mb(mb_old_funs) %>% mutate(type = "old")
tidy_mb_new <- tidy_mb(mb_new_funs) %>% mutate(type = "new")

mb_combined <- bind_rows(tidy_mb_old, tidy_mb_new)

# I think this is in microseconds? Unsure.
mb_combined %>%
  group_by(type,expr) %>%
  summarise_at(.vars = vars(time),
               .funs = funs(min,mean,median,max)) %>%
  arrange(expr) %>%
  knitr::kable()
```

| type | expr                |        min|         mean|       median|        max|
|:-----|:--------------------|----------:|------------:|------------:|----------:|
| new  | miss\_case\_pct     |     117943|     184379.5|     178148.0|     402108|
| old  | miss\_case\_pct     |     106193|     177510.1|     180681.0|     266083|
| new  | miss\_case\_prop    |     117773|     180140.1|     180333.5|     361126|
| old  | miss\_case\_prop    |     110790|     181768.4|     180751.0|     303183|
| new  | miss\_case\_summary |    8419355|    9754482.1|    9321849.5|   13886879|
| old  | miss\_case\_summary |  148829857|  173563729.5|  171862037.5|  242414116|
| new  | miss\_case\_table   |    5346805|    6507095.1|    5914116.5|   11927311|
| old  | miss\_case\_table   |  114623875|  129453846.5|  128055961.5|  199175146|
| new  | miss\_prop\_summary |    5833181|    6983529.0|    6496031.0|   11128861|
| old  | miss\_prop\_summary |    5451496|    7257661.4|    6552096.5|   19571237|
| new  | miss\_var\_pct      |     592197|     745296.5|     694597.5|    2849866|
| old  | miss\_var\_pct      |     575854|    1130419.0|     694072.5|   37490316|
| new  | miss\_var\_prop     |     589345|     737217.7|     673482.0|    2695309|
| old  | miss\_var\_prop     |     584099|     727344.2|     687114.5|    2647324|
| new  | miss\_var\_run      |    6059785|    7127802.0|    6559186.5|   14579901|
| old  | miss\_var\_run      |    5585718|    7049816.8|    6612873.0|   13353024|
| new  | miss\_var\_span     |   19837499|   24554941.6|   22591093.0|   95149500|
| old  | miss\_var\_span     |   19753235|   23586619.1|   22935410.5|   64614291|
| new  | miss\_var\_summary  |    4238269|    4978811.9|    4568662.0|    9697506|
| old  | miss\_var\_summary  |    4031169|    5054523.3|    4737219.5|    8254638|
| new  | miss\_var\_table    |    7108722|    8238089.9|    7847992.5|   12146477|
| old  | miss\_var\_table    |    6919870|    8810073.9|    7883966.5|   50184873|
| new  | gg\_miss\_case      |   14480615|   17156486.4|   16442428.0|   23802507|
| old  | gg\_miss\_case      |  142392590|  179270247.4|  179722623.0|  249276889|
| new  | gg\_miss\_fct       |   85789881|   96275362.5|   95870843.5|  164521438|
| old  | gg\_miss\_fct       |   86559868|   99718903.4|   97457500.5|  177047262|
| new  | gg\_miss\_var       |   10884530|   13189968.1|   11967101.5|   85998486|
| old  | gg\_miss\_var       |   10038797|   12704118.9|   12072091.5|   38914761|
| new  | gg\_miss\_span      |   27039057|   30972826.0|   30713547.0|   37507724|
| old  | gg\_miss\_span      |   25270603|   32757441.6|   30862518.5|  145657585|

``` r

library(ggplot2)
mb_combined %>%
  ggplot(aes(x = expr,
             y = time,
             colour = type)) +
  geom_boxplot() +
  coord_flip()
```

![](https://i.imgur.com/eu36sZW.png)

``` r

# let's focus on the ones that have a more noticeable difference
mb_combined %>%
  dplyr::filter(expr %in% c("gg_miss_case",
                     "miss_var_table",
                     "miss_var_summary",
                     "miss_var_span",
                     "miss_case_table",
                     "miss_case_summary")) %>%
  ggplot(aes(x = expr,
             y = time,
             colour = type)) +
  geom_boxplot() +
  coord_flip()
```

![](https://i.imgur.com/zDPbLwC.png)

**Comment by njtierney** [comment]
For the moment this issue is addressed, but the c++ branch will make everything faster once again. I will open up another issue for that comparison when we come to it.

**Comment by romainfrancois** [comment]
yep. Let's just not make a huge "C++ branch" though. Once we have the initial one merged in, we can make small branches that are easily reviewable and mergeable ...

**Comment by njtierney** [comment]
Yup, I agree - it is ideal to have a nice compact branch with incremental changes at each pull request :)
